minio/minio
# Support persistent queue store for loggers

## Description

This commit adds the queue store persistence support for MinIO loggers. This enables persisting the logs to the configured queue store which will be replayed by the store replay mechanism. This is similar to the queue store support for bucket notification targets.

This will make sure the logs will not be missed when the logger targets are down. The logs will be persisted in such cases and will be replayed once the target is up.

## Motivation and Context

To persist the logs so that no events will be missed when the target is down. Those interim events will be replayed back when the target is back online.

## How to test this PR?

- Setup webhook logger / audit webhook / audit kafka
- Enable the persistent store in the loggers. For example - `mc admin config set myminio audit_webhook enable=on endpoint=http://localhost:8080/ queue_size=100000 queue_dir=/tmp/webhook-logger`
- Bring down the target
- Generate logs - any s3 operations.
- Check the queue_dir - `mc tree --files /tmp/webhook-logger` - you should be seeing few log files persisted.
- Bring up the target
- After few secs, the events should have been replayed to the target and the queue store should be empty

(NOTE: For configuring kafka for testing, you can refer https://kafka.apache.org/quickstart)

## Types of changes

## Checklist:
